Minify js and css files and add them to build folder #5487

meltuhamy opened the pull request:
- follows on from #5593
- use npm for dependency on closure compiler and clean-css
- remove unneeded commas

Fixes #5487.

yurydelendik commented on make.js:
```diff
@@ -168,20 +168,20 @@ target.components = function() {
     'web/images/annotation-*.svg',
     'web/images/loading-icon.gif',
     'web/images/shadow.png',
-    'web/images/texture.png',
```
unrelated changes (here and below), please revert.

meltuhamy commented:
Hi @yurydelendik I've removed unnecessary changes as per your suggestion, and also fixed an issue which was causing the minified viewer not to display properly.  

A `minified` folder is now produced that contains all the needed minified files and also a fixed `viewer.html` that shows the preview pdf without errors.

mzgol commented:
Could minified files be put in the same directory as the built un-minified ones? Some tools assume that will be the case and they just swap `.js` to `.min.js`.

meltuhamy commented:
@mzgol thanks - my latest push now does this - however, there is a subtle difference: `pdf_viewer.min.js` *includes* `pdf.js`, while `pdf_viewer.js` *does not include* `pdf.js`.

The `dist` directory structure looks like this: 

```
dist
|   bower.json
|   LICENSE
|   package.json
|   README.md
|
+---build
|       pdf.combined.js
|       pdf.combined.min.js
|       pdf.js
|       pdf.min.js
|       pdf.worker.js
|       pdf.worker.min.js
|
+---cmaps
|       *.bcmap
|       LICENSE
|
\---web
    |   compatibility.js
    |   compatibility.min.js
    |   pdf_viewer.css
    |   pdf_viewer.js
    |   pdf_viewer.min.js
    |   viewer.min.css
    |
    \---images
            *.svg
            *.png

```

Snuffleupagus commented on make.js:
```diff
@@ -637,47 +645,78 @@ function cleanupCSSSource(file) {
+    var compilerDirPath = path.dirname(compilerPackagePath);
+    var compilerPath = path.join(compilerDirPath, 'compiler.jar');
+  } catch (e) {
+    console.log('google-closure-compiler is not installed. Run "npm install" ' +
```
Is there reason that you're using `console.log` instead of `echo` here (and below)?

meltuhamy replied:
Not really, I'm just used to console.log - happy to change to echo for consistency.

meltuhamy commented:
Code changed and ready for review again :)

mzgol commented:
> @mzgol thanks - my latest push now does this - however, there is a subtle difference: pdf_viewer.min.js includes pdf.js, while pdf_viewer.js does not include pdf.js.

Why is that necessary? Ideally `FILENAME.js` and `FILENAME.min.js` should differ only in that the latter is minified...

meltuhamy commented:
I agree but that would mean having 1 extra script tag so potentially 3 script requests would be made for the viewer to load. Also, there is no use having the viewer code without pdfjs loaded - ie the two files will always be required together. If a consumer needs pdfjs separately, that is also provided on its own in the build folder. 

The alternative is to provide the viewer code minified separately, and a combined viewer+pdfjs file in another file if anyone wishes to use that. 

Let me know what you think.

mzgol commented:
> The alternative is to provide the viewer code minified separately, and a combined viewer+pdfjs file in another file if anyone wishes to use that.

Note that I'm not a maintainer of this module so all I say is only my personal opinion and the maintainers might decide otherwise. So you might want to wait for an official opinion before changing anything.

That said, I think that would be great to have a combined viewer+pdfjs file and then its minified version. This would go in line with conventions other libraries take so it would reduce confusion in developers using pdf.js.

yurydelendik commented:
Currently our dist, web and release packaging is a little bit messy, but I agree that we shall match convention @mzgol mentioned above. Let's think if can / cannot package minified files in some cases without original source

meltuhamy commented:
@yurydelendik I've updated the branch, it should now follow the convention as suggested by @mgol 

The output now looks like the following:
```
+---minified
|   |   LICENSE
|   |   
|   +---build
|   |       pdf.combined.js
|   |       pdf.combined.min.js
|   |       pdf.js
|   |       pdf.min.js
|   |       pdf.worker.js
|   |       pdf.worker.min.js
|   |       
|   \---web
|       |   compatibility.js
|       |   compatibility.min.js
|       |   compressed.tracemonkey-pldi-09.pdf
|       |   debugger.js
|       |   l10n.js
|       |   pdf_viewer.min.js
|       |   viewer.css
|       |   viewer.html
|       |   viewer.js
|       |   viewer.min.css
|       |   viewer.min.js
|       |   
|       +---cmaps
|       |       *.bcmap
|       |     
|       +---images
|       |       *
|       |       
|       \---locale
|       |       *
```

meltuhamy commented:
What are your thoughts on packaging `preview.html` with the rest of the files? It would be good to provide the whole viewer as a minified application.

timvandermeij commented:
This patch needs a rebase after the UglifyJS patch that landed today. Additionally, please squash the commits into one afterwards (see https://github.com/mozilla/pdf.js/wiki/Squashing-Commits).

meltuhamy commented:
Will update this, but what are your thoughts on packaging `index.html` with the minified files? This means there is a demo that uses the minified files and also allows people to consume the reader as a whole (Eg. By putting it into an iframe) a lot easier. 

Let me know your thoughts.

timvandermeij commented:
We have `node make minified` to get the (minified) library and the viewer. The NPM package however should only contain the (minified) library. We can always add the viewer in a follow-up patch, but a while ago it was decided to not distribute the viewer via NPM because most people don't want to use the entire viewer (as it's specific to the Firefox use-case) and only make use of the component-based approach. On top of that, we ask developers to alter the default viewer to make it different from the viewer in Firefox, to not confuse users. If we ship the default viewer, this is less likely to happen.

mgol commented:
> The NPM package however should only contain the minified library.

I think it should contain both minified & unminified libraries. That's a pattern most libraries follows and it tremendously helps in debugging various issues. The only difference between the minified & unminified version should be the minification itself.

Snuffleupagus commented on make.js:
```diff
@@ -307,11 +309,22 @@ target.dist = function() {
   mkdir('-p', DIST_DIR + 'web/');
   cp('-R', [
     COMPONENTS_DIR + '*',
+    GENERIC_DIR + 'web/viewer.js',
```
Please note that the *only* files that should be present in the `dist` are the ones listed here: https://github.com/mozilla/pdfjs-dist/tree/master/web.
Providing minified versions of those files should be OK, but `viewer.{js, css}` should *not* be included!

meltuhamy replied:
I think you're right, it's maybe not in the scope of this PR, but I also think it would be good to publish a minified viewer. 

The use case I have in mind is if we want to export a generic viewer that can be included "out of the box" in other projects. Perhaps this is not the scope of the pdf.js project (it could be extracted into its own project that uses pdf.js), but then perhaps the whole viewer should not be included in this repo in that case. I appreciate it's quite a big and perhaps unnecessary change but if we're saying that the viewer shouldn't be published then it might make sense to put the viewer into its own repo and not version it here. 

I guess my question is: what is the role of the viewer in this repo? If it is a product that is bring released then maybe there's no harm on releasing a minified version.

timvandermeij replied:
The viewer is this repository is not intended to be used as-is by third-party solutions. Third-party viewers should always be different from the default viewer. Most people want a simple viewer, which they can achieve by using https://github.com/mozilla/pdf.js/tree/master/examples/components *or* by using parts of the default viewer. The default viewer is in this repository for two reasons: open-source development for the viewer in Firefox (and the Chrome extension) and third-party viewers might reuse parts of it.

It is really out of scope for this PR to include the viewer though.

meltuhamy replied:
Cool, thanks to the response. I'll patch the pr with the required changes  and squash it. Thanks

timvandermeij commented:
@mgol Sorry, indeed I meant that the package should contain the minified and unminified library, just not the viewer files.

meltuhamy committed:
Export minfied js and css files to pdf-dist

meltuhamy commented:
This is now done @timvandermeij

meltuhamy commented:
This is no longer needed as uglify js has been brought in :)
